mcs/class/System/System.Net.Sockets/SocketAsyncResult.cs

```diff
@@ -155,7 +155,11 @@ public void Complete ()
 
 			AsyncCallback callback = AsyncCallback;
 			if (callback != null) {
-				ThreadPool.UnsafeQueueUserWorkItem (_ => callback (this), null);
+				QueueUserWorkItemCallback workItemCallback = (object state) => {
```

Comment:
@marek-safar point out a race in my first commit, sth like
```
StaticFieldCallback = new  DelegateName( state => {
          ((SocketAsyncResult)state).AsyncCallback(state)
});

if ( this.AsyncCallback != null)
        ThreadPool.UnsafeQueueUserWorkItem(state =>
                StaticFieldCallback, this)
```
Is it possible anyone else touching SocketAsyncResult. AsyncCallback?